File: input/global_impl.zig
fn get_proc_addr(
    fp_get_instance_proc_addr: FpGetInstanceProcAddr,
    p_name: [*:0]const u8,
) MissingFunctionError!FpVoidFunction {
    return fp_get_instance_proc_addr(.null_handle, p_name) orelse return error.MissingFunction;
}

pub fn get_instance_proc_addr(
    self: GlobalCommands,
    instance: Instance,
    p_name: [*:0]const u8,
) MissingFunctionError!FpVoidFunction {
    return self.fp_get_instance_proc_addr(instance, p_name) orelse return error.MissingFunction;
}

pub const EnumerateInstanceVersionError = error{
    OutOfHostMemory,
    Unexpected,
};
pub fn enumerate_instance_version(
    self: GlobalCommands,
) EnumerateInstanceVersionError!Version {
    if (self.fp_enumerate_instance_version) |fp| {
        var version: Version = undefined;
        switch (fp(&version)) {
            .success => return version,
            .error_out_of_host_memory => return error.OutOfHostMemory,
            else => return error.Unexpected,
        }
    } else {
        return make_version(1, 0, 0);
    }
}

pub const CreateInstanceOrMissingFunctionError = CreateInstanceError || MissingFunctionError;
pub fn create_instance_to_commands(
    self: GlobalCommands,
    p_create_info: *const InstanceCreateInfo,
    p_allocator: ?*const AllocationCallbacks,
) CreateInstanceOrMissingFunctionError!InstanceCommands {
    const instance = try self.create_instance(p_create_info, p_allocator);
    return InstanceCommands.init(self, instance, p_create_info);
}

File: input/global_impl.rs
    pub unsafe fn enumerate_instance_version(&self) -> Result<vk::Version> {
        if let Some(fp) = self.fp_enumerate_instance_version {
            let mut p_api_version = MaybeUninit::<_>::uninit();
            let err = (fp)(p_api_version.as_mut_ptr());
            match err {
                vk::Result::SUCCESS => Ok(p_api_version.assume_init()),
                _ => Err(err),
            }
        } else {
            Ok(vk::Version::default())
        }
    }
    pub unsafe fn create_instance_commands(
        &self,
        p_create_info: &vk::InstanceCreateInfo,
        p_allocator: Option<&vk::AllocationCallbacks>,
    ) -> LoadResult<Instance> {
        let instance = self.create_instance(p_create_info, p_allocator)?;
        Instance::load(self, instance, p_create_info)
    }
